BUG: regression on Windows after ARPACK rewrite

There are 144 new failures on `main` after gh-22748 and/or gh-23274 (still need to check which one, building is slow). They are visible both on gh-23133 on `win_arm64` and locally for me on a conda-forge like build using `-Duse-g77-abi` and `clang-cl` on x86-64.

All the failures are for complex arrays, in the `test_hermitian` and `test_complex_nonsymmetric_modes` tests. Àll the numerical values are way off. Example failure:

```
scipy\sparse\linalg\_eigen\arpack\tests\test_arpack.py:137: AssertionError
____________________________________________________________________________ test_complex_nonsymmetric_modes[D1-D-LI-asarray-(0.1+0.1j)] ____________________________________________________________________________

actual = array([ 0.96375021-0.4039202j , -0.3598957 -0.78797294j]), desired = array([ 1.26434829-0.57929202j, -0.09499434-0.27649775j])
kw = {'atol': np.float64(4.440892098500626e-13), 'err_msg': 'error for eigs:general, typ=D, which=LI, sigma=(0.1+0.1j), mattype=asarray, OPpart=None, mode=normal', 'rtol': np.float64(4.440892098500626e-13)}

    def assert_allclose_cc(actual, desired, **kw):
        """Almost equal or complex conjugates almost equal"""
        try:
>           assert_allclose(actual, desired, **kw)
E           AssertionError: 
E           Not equal to tolerance rtol=4.44089e-13, atol=4.44089e-13
E           error for eigs:general, typ=D, which=LI, sigma=(0.1+0.1j), mattype=asarray, OPpart=None, mode=normal
E           Mismatched elements: 2 / 2 (100%)
E           Max absolute difference among violations: 0.57600313
E           Max relative difference among violations: 1.97017799
E            ACTUAL: array([ 0.96375 -0.40392j , -0.359896-0.787973j])
E            DESIRED: array([ 1.264348-0.579292j, -0.094994-0.276498j])

actual     = array([ 0.96375021-0.4039202j , -0.3598957 -0.78797294j])
desired    = array([ 1.26434829-0.57929202j, -0.09499434-0.27649775j])
kw         = {'atol': np.float64(4.440892098500626e-13), 'err_msg': 'error for eigs:general, typ=D, which=LI, sigma=(0.1+0.1j), mattype=asarray, OPpart=None, mode=normal', 'rtol': np.float64(4.440892098500626e-13)}

scipy\sparse\linalg\_eigen\arpack\tests\test_arpack.py:135: AssertionError

During handling of the above exception, another exception occurred:

D = <gen-cmplx-nonsym>, typ = 'D', which = 'LI', mattype = <built-in function asarray>, sigma = (0.1+0.1j)

    @pytest.mark.parametrize("sigma", [None, 0.1, 0.1 + 0.1j])
    @pytest.mark.parametrize("mattype", [csr_array, aslinearoperator, np.asarray])
    @pytest.mark.parametrize("which", ['LM', 'LR', 'LI'])
    @pytest.mark.parametrize("typ", ['F', 'D'])
    @pytest.mark.parametrize("D", NonSymmetricParams().complex_test_cases)
    def test_complex_nonsymmetric_modes(D, typ, which, mattype, sigma):
        rng = np.random.default_rng(1749533536274527)
        k = 2
>       eval_evec(False, D, typ, k, which, None, sigma, mattype, rng=rng)

D          = <gen-cmplx-nonsym>
k          = 2
mattype    = <built-in function asarray>
rng        = Generator(PCG64) at 0x2087ADBB140
sigma      = (0.1+0.1j)
typ        = 'D'
which      = 'LI'

scipy\sparse\linalg\_eigen\arpack\tests\test_arpack.py:467: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
scipy\sparse\linalg\_eigen\arpack\tests\test_arpack.py:273: in eval_evec
    assert_allclose_cc(eigenvalues, exact_eval, rtol=rtol, atol=atol, err_msg=err)
        OPpart     = None
        a          = array([[0.48870811+0.17661709j, 0.67635185+0.87407565j,
        0.59099644+0.82099736j, 0.25166088+0.97449869j,
      ...967j,
        0.80909383+0.01877711j, 0.98349106+0.498656j  ,
        0.25772843+0.91582274j, 0.73846334+0.55299526j]])
        ac         = array([[0.48870811+0.17661709j, 0.67635185+0.87407565j,
        0.59099644+0.82099736j, 0.25166088+0.97449869j,
      ...967j,
        0.80909383+0.01877711j, 0.98349106+0.498656j  ,
        0.25772843+0.91582274j, 0.73846334+0.55299526j]])
        atol       = np.float64(4.440892098500626e-13)
        b          = array([[2.52668095+0.j, 1.76407337+0.j, 1.6089164 +0.j, 1.02004337+0.j,
        2.55297852+0.j, 1.0468024 +0.j],
     ...0.j],
       [1.0468024 +0.j, 1.40759039+0.j, 1.88324583+0.j, 0.89514232+0.j,
        1.49642396+0.j, 2.00774384+0.j]])
        bc         = array([[2.52668095+0.j, 1.76407337+0.j, 1.6089164 +0.j, 1.02004337+0.j,
        2.55297852+0.j, 1.0468024 +0.j],
     ...0.j],
       [1.0468024 +0.j, 1.40759039+0.j, 1.88324583+0.j, 0.89514232+0.j,
        1.49642396+0.j, 2.00774384+0.j]])
        check_evecs = False
        d          = <gen-cmplx-nonsym>
        eigenvalues = array([ 0.96375021-0.4039202j , -0.3598957 -0.78797294j])
        eigs_func  = <function eigs at 0x00000207D10C68E0>
        err        = 'error for eigs:general, typ=D, which=LI, sigma=(0.1+0.1j), mattype=asarray, OPpart=None, mode=normal'
        evec       = array([[ 0.09866109-0.41058851j, -0.29003858-0.33536766j],
       [ 0.52141844+1.00314864j,  0.24689559+0.55489034j],
...
       [-0.31918192-0.25736273j, -0.55203096+0.12759985j],
       [-0.43002563-0.22698925j,  0.27734816+0.01445704j]])
        exact_eval = array([ 1.26434829-0.57929202j, -0.09499434-0.27649775j])
        general    = True
        ind        = array([1, 0])
        k          = 2
        kwargs     = {'OPpart': None, 'rng': Generator(PCG64) at 0x2087ADBB140, 'sigma': (0.1+0.1j), 'tol': 0, ...}
        mattype    = <built-in function asarray>
        mode       = 'normal'
        ntries     = 5
        rng        = Generator(PCG64) at 0x2087ADBB140
        rtol       = np.float64(4.440892098500626e-13)
        sigma      = (0.1+0.1j)
        symmetric  = False
        typ        = 'D'
        v0         = None
        which      = 'LI'
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

actual = array([ 0.96375021-0.4039202j , -0.3598957 -0.78797294j]), desired = array([ 1.26434829-0.57929202j, -0.09499434-0.27649775j])
kw = {'atol': np.float64(4.440892098500626e-13), 'err_msg': 'error for eigs:general, typ=D, which=LI, sigma=(0.1+0.1j), mattype=asarray, OPpart=None, mode=normal', 'rtol': np.float64(4.440892098500626e-13)}

    def assert_allclose_cc(actual, desired, **kw):
        """Almost equal or complex conjugates almost equal"""
        try:
            assert_allclose(actual, desired, **kw)
        except AssertionError:
>           assert_allclose(actual, conj(desired), **kw)
E           AssertionError: 
E           Not equal to tolerance rtol=4.44089e-13, atol=4.44089e-13
E           error for eigs:general, typ=D, which=LI, sigma=(0.1+0.1j), mattype=asarray, OPpart=None, mode=normal
E           Mismatched elements: 2 / 2 (100%)
E           Max absolute difference among violations: 1.09693691
E           Max relative difference among violations: 3.75199519
E            ACTUAL: array([ 0.96375 -0.40392j , -0.359896-0.787973j])
E            DESIRED: array([ 1.264348+0.579292j, -0.094994+0.276498j])

actual     = array([ 0.96375021-0.4039202j , -0.3598957 -0.78797294j])
desired    = array([ 1.26434829-0.57929202j, -0.09499434-0.27649775j])
kw         = {'atol': np.float64(4.440892098500626e-13), 'err_msg': 'error for eigs:general, typ=D, which=LI, sigma=(0.1+0.1j), mattype=asarray, OPpart=None, mode=normal', 'rtol': np.float64(4.440892098500626e-13)}

scipy\sparse\linalg\_eigen\arpack\tests\test_arpack.py:137: AssertionError
```

It looks like gh-23274 used raw BLAS signatures and fails to use the CBLAS interface, which explains the problem. It should use `npy_cblas.h`, as documented in gh-21130.

--- a/scipy/sparse/linalg/_eigen/arpack/arnaud/src/arnaud_n_double_complex.c
+++ b/scipy/sparse/linalg/_eigen/arpack/arnaud/src/arnaud_n_double_complex.c
@@ -364,7 +364,8 @@ ARNAUD_zneupd(struct ARNAUD_state_d *V, int rvec, int howmny, int* select,
         {
 #if defined(_MSC_VER)
             // Complex division is not supported in MSVC, multiply with reciprocal
-            temp = _Cmulcr(conj(workl[iheig + k]), 1.0 / cabs(workl[iheig + k]));
+            double mag_sq = pow(cabs(workl[iheig + k]), 2.0);
+            temp = _Cmulcr(conj(workl[iheig + k]), 1.0 / mag_sq);
             workl[ihbds + k] = _Cmulcc(_Cmulcc(workl[ihbds + k], temp), temp);
 #else
             temp = workl[iheig + k];
@@ -386,7 +387,8 @@ ARNAUD_zneupd(struct ARNAUD_state_d *V, int rvec, int howmny, int* select,
         {
 #if defined(_MSC_VER)
             // Complex division is not supported in MSVC
-            temp = _Cmulcr(conj(workl[iheig + k]), 1.0 / cabs(workl[iheig + k]));
+            double mag_sq = pow(cabs(workl[iheig + k]), 2.0);
+            temp = _Cmulcr(conj(workl[iheig + k]), 1.0 / mag_sq);
             d[k] = ARNAUD_cplx(creal(temp) + creal(sigma), cimag(temp) + cimag(sigma));
 #else
             d[k] = 1.0 / workl[iheig + k] + sigma;
@@ -415,7 +417,8 @@ ARNAUD_zneupd(struct ARNAUD_state_d *V, int rvec, int howmny, int* select,
             {
 #if defined(_MSC_VER)
                 // Complex division is not supported in MSVC
-                temp = _Cmulcr(conj(workl[iheig + j]), 1.0 / cabs(workl[iheig + j]));
+                double mag_sq = pow(cabs(workl[iheig + j]), 2.0);
+                temp = _Cmulcr(conj(workl[iheig + j]), 1.0 / mag_sq);
                 workev[j] = _Cmulcc(workl[invsub + j*ldq + V->ncv], temp);
 #else
                 workev[j] = workl[invsub + j*ldq + V->ncv] / workl[iheig+j];
@@ -1312,10 +1315,6 @@ znapps(int n, int* kev, int np, ARNAUD_CPLX_TYPE* shift, ARNAUD_CPLX_TYPE* v,
     double tmp_dbl;
     ARNAUD_CPLX_TYPE f, g, h11, h21, sigma, s, s2, r, t, tmp_cplx;
 
-#if defined(_MSC_VER)
-    ARNAUD_CPLX_TYPE tmp_cplx2;
-#endif
-
     ARNAUD_CPLX_TYPE cdbl1 = ARNAUD_cplx(1.0, 0.0);
     ARNAUD_CPLX_TYPE cdbl0 = ARNAUD_cplx(0.0, 0.0);
 
@@ -1513,7 +1512,8 @@ zneigh(double* rnorm, int n, ARNAUD_CPLX_TYPE* h, int ldh, ARNAUD_CPLX_TYPE* rit
     int select[1] = { 0 };
     int int1 = 1, j;
     double temp;
-    ARNAUD_CPLX_TYPE vl[1] = { 0.0 };
+    ARNAUD_CPLX_TYPE vl[1];
+    vl[0] = ARNAUD_cplx(0.0, 0.0);
     ARNAUD_CPLX_TYPE c1 = ARNAUD_cplx(1.0, 0.0), c0 = ARNAUD_cplx(0.0, 0.0);
 
     //  1. Compute the eigenvalues, the last components of the

--- a/scipy/sparse/linalg/_eigen/arpack/arnaud/src/arnaud_n_single_complex.c
+++ b/scipy/sparse/linalg/_eigen/arpack/arnaud/src/arnaud_n_single_complex.c
@@ -364,7 +364,8 @@ ARNAUD_cneupd(struct ARNAUD_state_s *V, int rvec, int howmny, int* select,
         {
 #if defined(_MSC_VER)
             // Complex division is not supported in MSVC, multiply with reciprocal
-            temp = _FCmulcr(conjf(workl[iheig + k]), 1.0 / cabsf(workl[iheig + k]));
+            float mag_sq = powf(cabsf(workl[iheig + k]), 2.0);
+            temp = _FCmulcr(conjf(workl[iheig + k]), 1.0 / mag_sq);
             workl[ihbds + k] = _FCmulcc(_FCmulcc(workl[ihbds + k], temp), temp);
 #else
             temp = workl[iheig + k];
@@ -386,7 +387,8 @@ ARNAUD_cneupd(struct ARNAUD_state_s *V, int rvec, int howmny, int* select,
         {
 #if defined(_MSC_VER)
             // Complex division is not supported in MSVC
-            temp = _FCmulcr(conjf(workl[iheig + k]), 1.0 / cabsf(workl[iheig + k]));
+            float mag_sq = powf(cabsf(workl[iheig + k]), 2.0);
+            temp = _FCmulcr(conjf(workl[iheig + k]), 1.0 / mag_sq);
             d[k] = ARNAUD_cplxf(crealf(temp) + crealf(sigma), cimagf(temp) + cimagf(sigma));
 #else
             d[k] = 1.0 / workl[iheig + k] + sigma;
@@ -415,7 +417,8 @@ ARNAUD_cneupd(struct ARNAUD_state_s *V, int rvec, int howmny, int* select,
             {
 #if defined(_MSC_VER)
                 // Complex division is not supported in MSVC
-                temp = _FCmulcr(conjf(workl[iheig + j]), 1.0 / cabsf(workl[iheig + j]));
+                float mag_sq = powf(cabsf(workl[iheig + j]), 2.0);
+                temp = _FCmulcr(conjf(workl[iheig + j]), 1.0 / mag_sq);
                 workev[j] = _FCmulcc(workl[invsub + j*ldq + V->ncv], temp);
 #else
                 workev[j] = workl[invsub + j*ldq + V->ncv] / workl[iheig+j];
@@ -1312,10 +1315,6 @@ cnapps(int n, int* kev, int np, ARNAUD_CPLXF_TYPE* shift, ARNAUD_CPLXF_TYPE* v,
     float tmp_dbl;
     ARNAUD_CPLXF_TYPE f, g, h11, h21, sigma, s, s2, r, t, tmp_cplx;
 
-    #if defined(_MSC_VER)
-    ARNAUD_CPLXF_TYPE tmp_cplx2;
-    #endif
-
     ARNAUD_CPLXF_TYPE cdbl1 = ARNAUD_cplxf(1.0, 0.0);
     ARNAUD_CPLXF_TYPE cdbl0 = ARNAUD_cplxf(0.0, 0.0);
 
@@ -1513,7 +1512,8 @@ cneigh(float* rnorm, int n, ARNAUD_CPLXF_TYPE* h, int ldh, ARNAUD_CPLXF_TYPE* ri
     int select[1] = { 0 };
     int int1 = 1, j;
     float temp;
-    ARNAUD_CPLXF_TYPE vl[1] = { 0.0 };
+    ARNAUD_CPLXF_TYPE vl[1];
+    vl[0] = ARNAUD_cplxf(0.0, 0.0);
     ARNAUD_CPLXF_TYPE c1 = ARNAUD_cplxf(1.0, 0.0), c0 = ARNAUD_cplxf(0.0, 0.0);
 
     //  1. Compute the eigenvalues, the last components of the

